numpy/testing/nose_tools/utils.py

```diff
@@ -394,14 +394,17 @@ def assert_equal(actual, desired, err_msg='', verbose=True):
         isdesnat = isnat(desired)
         isactnat = isnat(actual)
         dtypes_match = array(desired).dtype.type == array(actual).dtype.type
-        if isdesnat and isactnat and dtypes_match:
+        if isdesnat and isactnat:
             # If both are NaT (and have the same dtype -- datetime or
             # timedelta) they are considered equal.
-            return
+            if dtypes_match:
+                return
+            else:
+                raise AssertionError(msg)
```

Comment:
This is at best a workaround. It doesn't solve the underlying problem, so the following still has the warning-swallowing problems:
```
assert_equal(np.void(b'test'), np.datetime64('2017'))
```

Reply:
It may be incomplete, but I don't think it is a workaround, differing types should be handled with the NaT stuff, not passed on to "==". For the rest, I'm thinking that we should not intercept any of the warning exceptions, just raise all of them. It is only in development mode that they are raised, and those errors should also be fixed in development mode.

Reply:
I see two approaches here for the `==` fallback,

- Suppress deprecation warnings, assert_equal should do its thing regardless.
- Let deprecation warnings raise without conversion to `AssertionError`

The first can lead to unexpected changes when deprecations expire, the second will generally require fixes or rewrites that could conceivably be difficult before the deprecation expires.